### DISCUSSION TITLE
Decouple validator from AaC spec

### DISCUSSION THREAD

**Discussion by cvchaparro** [DevOps-MBSE/AaC#56] [Archive] [answered]
For continued discussion on the comment in PR #47.

> It looks like a lot of the AaC structure is becoming hard coded into the validator. How can we avoid that and use the AaC specification of itself to drive the validation? The test would be to change the AaC spec and successfully validate itself without having to change anything in the validator.

As mentioned in that comment (and my response to it) we really need a way to decouple the validator from the spec enough to where the validator can automatically account for changes to the spec without the need for manual intervention. In Lisp, I'd do this with macros and generate classes based on each type read in from the spec. Unfortunately, I still don't know enough python to know how to achieve the same end. That said, Alex mentioned that we could achieve a similar thing in Python using Jinja2 templates, for example. Seems like it could at least be worth a try.

That said, here are a few things I think need to be addressed to allow us to decouple the validator and the spec:

1. The "magic strings" referencing DSL names such as [model](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L312), [Behavior](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L313), [Field](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L239) [in](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L291) [several](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L314) [places](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L319), etc.
2. Another place we have some hard-coded knowledge of the spec is when [validating](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L136) [cross-references](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L145-L147).
3. Also, I know there is some hard-coded knowledge is in some of the [function names](https://github.com/jondavid-black/AaC/blob/46bb74c14a447aa7d6c03e9e229c6051bb6c3dc4/src/aac/validator.py#L28-L32) for validating specific parts of the spec.

**Comment by jondavid-black** [comment]
The primary concern I have is the existence of a validation function for each root type.  I'd like for "data" and "enum" to be the only root types we permit special knowledge of in validation if possible.  I want to be able to create a new root type through extension and not have to change the core validator in any way.

Cross references are a real challenge.  Maybe we need a different way to indicate type references in the spec so the validator can detect the need to perform cross-reference validation without "hard coded" knowledge.

**Reply by cvchaparro** [reply]
I agree with the assessment that we should not have a separate function for each type of model to validate. In fact, my intention was/is to remove those extra functions by replacing them with a function that achieved the same task. That is the main thing I was getting at with [this TODO comment](https://github.com/jondavid-black/AaC/blob/5a7824b4c6629b9642286533a13b85bbaefdb4d2/src/aac/validator.py#L4). I wanted to combine all the `get_all_*_errors` functions into `get_all_errors` (which was just recently renamed to `validate_and_get_errors`) which is why I was trying to structure all the functions as identically as possible. In this way, I'd be able to extract the pattern into a single function that handles all of the different model types without having to delegate to a separate function. That said, there are enough minor differences in checking (currently) the properties/structure of each model that I wasn't able to complete that.

Of course, I've also been brainstorming some ways to handle this more generically and I think my initial attempt (if I were to do this) will be to (programmatically) look through the fields of the model and, based on the data type of each one, recursively validate that way. I think this would allow me to rely on the knowledge that data models have fields without having to know beforehand what exactly the field's type is.

As far as cross-referencing, I'll have to think more on it because that was a part that I didn't (and still don't) fully understand what was going on.

With all of this in mind, if I work on this (which is my assumption unless you don't want me to do it - I like to finish what I start) I'll need to start prototyping these ideas because right now they're a bit nebulous in my mind.

**Comment by Coffee2Bits** [comment, marked as answer]
I think we can retire this conversation in lieu of #125 the validator 2.0.

**Comment by lizzcondrey** [comment]
This has been marked resolved, has an answer, and has the follow-on solution discussion linked. Closing this discussion.
